Compatibility with blog 1.0

tractorcow opened the pull request:
This change introduces:
- Default theme included in base templates (can be overridden in user templates) based on https://github.com/micmania1/simple_blog
- Inclusion of legacy migration classes for old BlogTree, BlogHolder and BlogEntry pages, as well as ArchiveWidget and TagCloudWidget.
- Migration task, which can be configured to be either manually run, or run automatically during dev/build (on by default)

micmania1 commented on code/model/Blog.php:
```diff
@@ -266,21 +266,21 @@ public function rss() {
-		$posts->setPageStart($start);
-
-		return $posts;
+		// Set page size
```
I think regardless of whether or not all results are returned, it should return PaginatedList for simplicity of the API.

tractorcow replied:
I guess

tractorcow replied:
I've made a PR to disable pagination within the list itself. https://github.com/silverstripe/silverstripe-framework/pull/3970

The existing code crashes if paging = 0 and num-items = 0, so I'll just fix that then.

micmania1 commented on templates/Layout/BlogPost.ss:
```diff
@@ -0,0 +1,17 @@
+<% require css(themes/simple_blog/css/blog.css) %>
```
The path needs updated. I think its fair to assume the folder name here rather than muddying the waters with PHP requirements.

tractorcow replied:
Ah yeah. We should use themedCSS helper anyway.

micmania1 replied:
I don't think themedCSS will work for modules.

tractorcow replied:
That's exactly what it's for. :) The second parameter is the module name.

micmania1 replied:
I never knew that :smile: I always just used <% require css %>

micmania1 commented on templates/widgets/BlogArchiveWidget.ss:
```diff
@@ -1,7 +1,12 @@
 <% if Archive %>
```
Should be changed to $Archive

micmania1 commented on templates/widgets/BlogArchiveWidget.ss:
```diff
@@ -1,7 +1,12 @@
 <% if Archive %>
 	<ul>
 		<% loop Archive %>
```
and here.

micmania1 commented on templates/Layout/Blog.ss:
```diff
@@ -0,0 +1,64 @@
+		<div class="content"><p>No Posts</p></div>
+	<% end_if %>
+
+	<% if PaginatedList.MoreThanOnePage %>
```
Have you tested this without any blog posts. (ie. the division by zero error we seen the other day).

tractorcow replied:
Yep, but i'll revert the API anyway. It's always ok to call non-existent methods in templates.

micmania1 commented on templates/Layout/Blog.ss:
```diff
@@ -0,0 +1,64 @@
+		<% end_if %>
+	</h1>
+
+	<% if PaginatedList %>
```
$PaginatedList

micmania1 commented on templates/Layout/Blog.ss:
```diff
@@ -0,0 +1,64 @@
+
+	<% if PaginatedList %>
+	<ul id="SearchResults">
+		<% loop PaginatedList %>
```
and here.

micmania1 commented on templates/Layout/Blog.ss:
```diff
@@ -0,0 +1,64 @@
+<% include BlogSideBar %>
+<div class="unit size3of4 lastUnit searchResults">
+	<h1>
+		<% if ArchiveYear %>
```
$ArchiveYear

micmania1 commented on templates/Layout/Blog.ss:
```diff
@@ -0,0 +1,64 @@
+	<h1>
+		<% if ArchiveYear %>
+			Archive: <% if ArchiveDay %>$ArchiveDate.Nice<% else_if ArchiveMonth %>$ArchiveDate.format("F, Y")<% else %>$ArchiveDate.format(Y)<% end_if %>
+		<% else_if CurrentTag %>
```
$CurrentTag

micmania1 commented on templates/Layout/Blog.ss:
```diff
@@ -0,0 +1,64 @@
+			Archive: <% if ArchiveDay %>$ArchiveDate.Nice<% else_if ArchiveMonth %>$ArchiveDate.format("F, Y")<% else %>$ArchiveDate.format(Y)<% end_if %>
+		<% else_if CurrentTag %>
+			Tag: $CurrentTag.Title
+		<% else_if CurrentCategory %>
```
etc.

micmania1 commented on templates/Includes/EntryMeta.ss:
```diff
@@ -0,0 +1,17 @@
+<p>
+	This entry was
+	<% if Categories %>
```
$Categories

micmania1 commented on templates/Includes/EntryMeta.ss:
```diff
@@ -0,0 +1,17 @@
+	This entry was
+	<% if Categories %>
+		posted in
+		<% loop Categories %>
```
and here.

micmania1 commented on templates/Includes/EntryMeta.ss:
```diff
@@ -0,0 +1,17 @@
+			<a href="$Link" title="$Title">$Title</a><% if not Last %>, <% else_if Up.Tags %>, <% else %> and<% end_if %>
+		<% end_loop %>
+	<% end_if %>
+	<% if Tags %>
```
$Tags

micmania1 commented on templates/Includes/EntryMeta.ss:
```diff
@@ -0,0 +1,17 @@
+	<% end_if %>
+	<% if Tags %>
+		tagged
+		<% loop Tags %>
```
and here.

micmania1 commented on code/compat/pages/BlogEntry.php:
```diff
@@ -0,0 +1,72 @@
+		return false;
+	}
+
+	public function up() {
```
Hooks should be provided in up() methods so that devs can upgrade bespoke code.

tractorcow replied:
Sure, that sounds doable.

tractorcow replied:
I've added the hook directly into the upgrade task below.

micmania1 commented on code/compat/pages/BlogEntry.php:
```diff
@@ -0,0 +1,72 @@
+	 * @return array Associative array of lowercase tag to native case tags
+	 * @deprecated since version 2.0
+	 */
+	public function TagNames() {
```
Can this be made private so that its not used outside of this class?

tractorcow replied:
That would defeat the purpose of deprecating a method, if you're going to immediately make it inaccessible. The idea is to provide a graceful upgrade process.

micmania1 replied:
Ok, just a bit confusing. To my module its new code that's being immediately added and deprecated, but to this blog module its old code that's being gracefully deprecated.

micmania1 commented on code/compat/pages/BlogEntry.php:
```diff
@@ -0,0 +1,72 @@
+	public function requireDefaultRecords() {
+		parent::requireDefaultRecords();
+
+		if(BlogMigrationTask::config()->run_during_dev_build) {
```
I don't think we should give the option to run via dev/build. The more decoupled from the core it is makes it easier to remove later.

tractorcow replied:
All of the code is in the same compat folder, so removing it later is as easy as deleting that folder.

micmania1 commented:
I've left some comments - mostly little things.

tractorcow commented:
I've now removed all templates from this, and updated the travis / unit tests.

Template PR i now at https://github.com/silverstripe/silverstripe-blog/pull/148

tractorcow commented:
Travis passes although scrutuniser failed.
